### PR TITLE
counter: add generic implementation and use it to replace StringCounter

### PR DESCRIPTION
**Pull request by tklauser**
The generic implementation will be used in a follow-up change to replace `StringCounter` instances tracking backends in `pkg/service` to use `counter.Counter[netip.Addr]`.

Keep IntCounter as is for now because for `IntCounter.ToBPFData` it requires the map elements to be ordered (i.e. `constraints.Ordered`), not just comparable.  This method is only used on instances of type `IntCounter` so far and making the map elements ordered would limit its use e.g. with `netip.Addr`.

### CODE REVIEW

File: pkg/counter/counter.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+// Counter tracks references for comparable .
+//
+// No thread safety is provided within this structure, the user is expected to
+// handle concurrent access to this structure if it is used from multiple
+// threads.
+type Counter[T comparable] map[T]int
+
+// Add increments the reference count for the specified key.
+func (c Counter[T]) Add(key T) (changed bool) {
```
Comment (aspsk) [inline pkg/counter/counter.go]: Won't it be more readable like this?
```
func (c Counter[T]) Add(key T) bool {
        value, exists := c[key]
        c[key] = value + 1
        return !exists
}
```
(for me `changed` sounds like the value was changed)

**Review by aspsk — APPROVED** [review]
Looks good, one comment about code, but your's variant looks fine as well

File: pkg/counter/counter.go (hunk 1 above)
Comment (tklauser) [reply pkg/counter/counter.go]: Agree, thanks for the suggestion.

**Review by ldelossa — APPROVED** [review]
Nice!

**Comment by tklauser** [comment]
/test-1.25-net-next

**Pull request merged**